src/store/sled.rs

```diff
@@ -18,193 +15,196 @@ use libsignal_service::{
     where
-        K: AsRef<str>,
-        IVec: From<V>,
+        T: AsRef<[u8]>,
+        K: AsRef<[u8]>,
+        V: Serialize,
     {
-        trace!("inserting {}", key.as_ref());
+        let encrypted_value = self.cipher.as_ref().map_or_else(
+            || serde_json::to_vec(&value).map_err(Error::from),
+            |c| c.encrypt_value(&value).map_err(Error::from),
+        )?;
         let _ = self
-            .db
-            .try_write()
-            .expect("poisoned mutex")
-            .insert(key.as_ref(), value)?;
-        Ok(())
-    }
-
-    fn insert_u32<S>(&self, key: S, value: u32) -> Result<(), Error>
-    where
-        S: AsRef<str>,
-    {
-        trace!("inserting u32 {}", key.as_ref());
-        self.db
-            .try_write()
-            .expect("poisoned mutex")
-            .insert(key.as_ref(), &value.to_le_bytes())?;
+            .tree(tree.as_ref())?
+            .insert(key.as_ref(), encrypted_value)?;
         Ok(())
     }
 
-    fn remove<S>(&self, key: S) -> Result<(), Error>
+    fn remove<T, K>(&self, tree: T, key: K) -> Result<bool, Error>
     where
-        S: AsRef<str>,
+        T: AsRef<[u8]>,
+        K: AsRef<[u8]>,
     {
-        trace!("removing {} from db", key.as_ref());
-        self.db
-            .try_write()
-            .expect("poisoned mutex")
-            .remove(key.as_ref())?;
-        Ok(())
-    }
-
-    fn prekey_key(&self, id: PreKeyId) -> String {
-        format!("prekey-{:09}", id)
-    }
-
-    fn signed_prekey_key(&self, id: SignedPreKeyId) -> String {
-        format!("signed-prekey-{:09}", id)
-    }
-
-    fn session_key(&self, addr: &ProtocolAddress) -> String {
-        format!("session-{}", addr)
-    }
-
-    fn session_prefix(&self, name: &str) -> String {
-        format!("session-{}.", name)
-    }
-
-    fn identity_key(&self, addr: &ProtocolAddress) -> String {
-        format!("identity-remote-{}", addr)
+        Ok(self.tree(tree)?.remove(key.as_ref())?.is_some())
     }
 
-    pub fn keys(&self) -> Result<(Vec<String>, Vec<String>), SignalProtocolError> {
-        let db = self.db.read().expect("poisoned mutex");
-        let global_keys = db
-            .iter()
-            .filter_map(|r| {
-                let (k, _) = r.ok()?;
-                Some(String::from_utf8_lossy(&k).to_string())
-            })
-            .collect();
-        let session_keys = db
-            .open_tree(SLED_TREE_SESSIONS)
-            .unwrap_or_else(|e| {
-                panic!("failed to open sessions tree: {}", e);
-            })
-            .iter()
-            .filter_map(|r| {
-                let (k, _) = r.ok()?;
-                Some(String::from_utf8_lossy(&k).to_string())
-            })
-            .collect();
-        Ok((global_keys, session_keys))
+    /// build a hashed messages thread key
+    fn messages_thread_key(&self, t: &Thread) -> String {
+        format!("{SLED_TREE_THREAD_PREFIX}{t:?}")
```

Comment:
This should hash at some point, right? Is it for hiding the thread id?

Reply:
Yes, I have a pending commit to deal with some of the keys.